src/detect-smb-version.c

```diff
@@ -0,0 +1,167 @@
+
+#include "detect-smb-version.h"
+#include "rust.h"
+
+#define BUFFER_NAME  "smb_version"
+#define KEYWORD_NAME "smb.version"
+#define KEYWORD_ID   DETECT_SMB_VERSION
+
+static int g_smb_version_list_id = 0;
+
+static void DetectSmbVersionFree(DetectEngineCtx *de_ctx, void *ptr)
+{
+    SCEnter();
+
+    SCLogDebug("smb_version: DetectSmbVersionFree");
+
+    if (ptr != NULL) {
+        rs_smb_version_free(ptr);
+    }
+    SCReturn;
+}
+
+/**
+ * \brief Creates a SigMatch for the "smb.version" keyword being sent as argument,
+ *        and appends it to the rs_smb_version_match Signature(s).
+ *
+ * \param de_ctx Pointer to the detection engine context.
+ * \param s      Pointer to signature for the current Signature being parsed
+ *               from the rules.
+ * \param arg    Pointer to the string holding the keyword value.
+ *
+ * \retval 0 on success, -1 on failure
+ */
+
+static int DetectSmbVersionSetup(DetectEngineCtx *de_ctx, Signature *s, const char *arg)
+{
+    SCLogDebug("smb_version: DetectSmbVersionSetup");
+
+    if (DetectSignatureSetAppProto(s, ALPROTO_SMB) < 0)
+        return -1;
+
+    if (arg == NULL) {
+        SCLogError("Error parsing smb.version option in signature, it needs a value");
+        return -1;
+    }
+
+    void *dod = rs_smb_version_parse(arg);
+    if (dod == NULL) {
+        SCLogError("Error parsing smb.version option in signature");
+        return -1;
+    }
+
+    if (DetectGetLastSMFromLists(s, DETECT_SMB_VERSION, -1)) {
+        SCLogError("Can't use 2 or more smb.version declarations in "
+                   "the same sig. Invalidating signature.");
+        return -1;
+    }
+
+    SigMatch *sm = SigMatchAlloc();
+
+    if (sm == NULL) {
+        DetectSmbVersionFree(de_ctx, dod);
+        return -1;
+    }
+
+    sm->type = DETECT_SMB_VERSION;
+    sm->ctx = dod;
+
+    SigMatchAppendSMToList(s, sm, g_smb_version_list_id);
+    return 0;
+}
+
+/**
+ * \brief App layer match function for the "smb.version" keyword.
+ *
+ * \param t       Pointer to the ThreadVars instance.
+ * \param det_ctx Pointer to the DetectEngineThreadCtx.
+ * \param f       Pointer to the flow.
+ * \param flags   Pointer to the flags indicating the flow direction.
+ * \param state   Pointer to the app layer state data.
+ * \param s       Pointer to the Signature instance.
+ * \param m       Pointer to the SigMatch.
+ *
+ * \retval 1 On Match.
+ * \retval 0 On no match.
+ */
+
+static int DetectSmbVersionMatchRust(DetectEngineThreadCtx *det_ctx, Flow *f, uint8_t flags,
+        void *state, void *txv, const Signature *s, const SigMatchCtx *m)
+{
+    SCEnter();
+
+    SCLogDebug("smb_version: DetectSmbVersionMatchRust");
+
+    int matchvalue = rs_smb_version_match(txv, (void *)m);
+
+    if (matchvalue != 1) {
+        SCLogDebug("rs_smb_version_match: didn't match");
+        SCReturnInt(0);
+    } else {
+        return matchvalue;
+        SCLogDebug("rs_smb_version_match: matched!");
+    }
+
+    SCReturnInt(1);
```

Comment:
dead code

Reply:
the SCReturnInt(); calls can be removed then?

Reply:
And `SCLogDebug("rs_smb_version_match: matched!");` that is right after `return` 

Right ?

Reply:
Oops, you are of course correct. Will have this updated in the next PR. Thanks!